res/controllers/Numark-Mixtrack-Pro-FX-scripts.js

```diff
@@ -0,0 +1,782 @@
+    });
+
+    this.pflButton = new components.Button({
+        type: components.Button.prototype.types.toggle,
+        midi: [0x90 + channel, 0x1B],
+        key: "pfl"
+    });
+
+    this.loadButton = new components.Button({
+        inKey: "LoadSelectedTrack"
+    });
+
+    this.volume = new components.Pot({
+        group: this.currentDeck,
+        inKey: "volume"
+    });
+
+    this.treble = new components.Pot({
+        group: "[EqualizerRack1_" + this.currentDeck + "_Effect1]",
+        inKey: "parameter3"
+    });
+
+    this.mid = new components.Pot({
+        group: "[EqualizerRack1_" + this.currentDeck + "_Effect1]",
+        inKey: "parameter2"
+    });
+
+    this.bass = new components.Pot({
+        group: "[EqualizerRack1_" + this.currentDeck + "_Effect1]",
+        inKey: "parameter1"
+    });
+
+    this.filter = new components.Pot({
+        group: "[QuickEffectRack1_" + this.currentDeck + "]",
+        inKey: "super1"
+    });
+
+    this.gain = new components.Pot({
+        inKey: "pregain"
+    });
+
+    this.pitch = new components.Pot({
+        inKey: "rate",
+        invert: true
+    });
+
+    this.padSection = new MixtrackProFX.PadSection(number);
+
+    this.shiftButton = new components.Button({
+        input: function(channel, control, value) {
+            // each shift button shifts both decks
+            // more consistent with the logic burned into hardware
+            if (this.isPress(channel, control, value)) {
+                MixtrackProFX.shift();
+            } else {
+                MixtrackProFX.unshift();
+            }
+        },
+        isPress: function(channel, control, value) {
+            if (value === 0x7F) {
+                return true;
+            }
+            return false;
+        }
+    });
+
+    this.loop = new components.Button({
+        outKey: "loop_enabled",
+        midi: [0x94 + channel, 0x40],
+        shift: function() {
+            this.inKey = "loop_in_goto";
+            this.input = components.Button.prototype.input;
+        },
+        unshift: function() {
+            this.input = function(channel, control, value, status, group) {
+                if (value !== 0x7F) {
+                    return;
+                }
+                if (engine.getValue(group, "loop_enabled") === 0) {
+                    script.triggerControl(group, "beatloop_activate");
+                } else {
+                    script.triggerControl(group, "beatlooproll_activate");
+                }
+            };
+        },
+        shiftControl: true,
+        sendShifted: true,
+        shiftOffset: 0x01
+    });
+
+    this.loopHalf = new components.Button({
+        midi: [0x94 + channel, 0x34],
+        shiftControl: true,
+        sendShifted: true,
+        shiftOffset: 0x02,
+        shift: function() {
+            this.inKey = "loop_in";
+            this.outKey = "loop_in";
+            this.disconnect();
+            this.connect();
+        },
```

Comment:
You could disconnect by some higher-order-function foo. 
create a global function:
```javascript
MixtrackProFX.createLayerMethod = function (key) {
    return function () {
        this.disconnect();
        this.inKey = key;
        this.outKey = key;
        this.connect();
    }
};
```
and then use it everywhere like this:
```suggestion
        shift: MixtrackProFX.createLayerMethod("loop_in").bind(this)
```

@Be-ing @Holzhaus what you think?

Reply:
I'd prefer to disconnect first, and reconnect afterwards. Other than that I like the current version. Let's try to keep "fancy JS stuff" to a minimum to keep stuff readable for newbies.

Reply:
I guess we can leave it out. though I wouldn't call it fancy. Its some basic functional programming and I think newbies should be able to wrap their head around it if we provide a gentle introduction in our wiki.

For now, I'm okay with leaving it out.